File: modules/screen-state/android/src/main/java/expo/modules/screenstate/ScreenStateModule.kt
package expo.modules.screenstate

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

// 화면 잠금(끄기)·해제를 JS로 이벤트 전달 — 세션 이탈 판정에서
// "폰 잠금(집중 인정)"과 "다른 앱 사용(차감)"을 구분하기 위한 모듈
class ScreenStateModule : Module() {
  private var receiver: BroadcastReceiver? = null

  private fun unregister() {
    receiver?.let {
      runCatching { appContext.reactContext?.unregisterReceiver(it) }
    }
    receiver = null
  }

  override fun definition() = ModuleDefinition {
    Name("ScreenState")

    Events("onScreenState")

    OnStartObserving {
      if (receiver == null) {
        receiver = object : BroadcastReceiver() {
          override fun onReceive(context: Context?, intent: Intent?) {
            val state = when (intent?.action) {
              Intent.ACTION_SCREEN_OFF -> "off"
              Intent.ACTION_SCREEN_ON -> "on"
              Intent.ACTION_USER_PRESENT -> "present"
              else -> return
            }
            sendEvent("onScreenState", mapOf("state" to state))
          }
        }
        val filter = IntentFilter().apply {
          addAction(Intent.ACTION_SCREEN_OFF)
          addAction(Intent.ACTION_SCREEN_ON)
          addAction(Intent.ACTION_USER_PRESENT)
        }
        appContext.reactContext?.registerReceiver(receiver, filter)
      }
    }

    OnStopObserving {
      unregister()
    }

    OnDestroy {
      unregister()
    }
  }
}

File: modules/screen-state/ios/ScreenStateModule.swift
import ExpoModulesCore

// iOS 잠금 감지: 기기에 암호가 설정된 경우 잠금 시 protectedDataWillBecomeUnavailable,
// 해제 시 protectedDataDidBecomeAvailable이 발생한다. (암호 미설정 기기는 이벤트 없음 → JS 폴백)
public class ScreenStateModule: Module {
  private var observers: [NSObjectProtocol] = []

  public func definition() -> ModuleDefinition {
    Name("ScreenState")

    Events("onScreenState")

    OnStartObserving {
      let lock = NotificationCenter.default.addObserver(
        forName: UIApplication.protectedDataWillBecomeUnavailableNotification,
        object: nil,
        queue: .main
      ) { [weak self] _ in
        self?.sendEvent("onScreenState", ["state": "off"])
      }
      let unlock = NotificationCenter.default.addObserver(
        forName: UIApplication.protectedDataDidBecomeAvailableNotification,
        object: nil,
        queue: .main
      ) { [weak self] _ in
        self?.sendEvent("onScreenState", ["state": "present"])
      }
      self.observers = [lock, unlock]
    }

    OnStopObserving {
      self.observers.forEach { NotificationCenter.default.removeObserver($0) }
      self.observers = []
    }
  }
}
